component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/duo/DuoAuthenticator.java

```diff
@@ -469,28 +537,51 @@ public List<Property> getConfigurationProperties() {
     protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
                                                  AuthenticationContext context) throws AuthenticationFailedException {
 
-        Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
         Map<String, String> duoParameters = getAuthenticatorConfig().getParameterMap();
+        String requestState = request.getParameter(DuoAuthenticatorConstants.DUO_STATE);
+        String requestDuoCode = request.getParameter(DuoAuthenticatorConstants.DUO_CODE);
+
         try {
-            String username = DuoWeb.verifyResponse(authenticatorProperties.get
-                            (DuoAuthenticatorConstants.INTEGRATION_KEY), authenticatorProperties.get
-                            (DuoAuthenticatorConstants.SECRET_KEY), context.getProperty
-                            (DuoAuthenticatorConstants.INTEGRATION_SECRET_KEY).toString(),
-                    request.getParameter(DuoAuthenticatorConstants.SIG_RESPONSE));
+            // Step 5: Validate state returned from Duo is the same as the one saved previously.
+            // If it isn't return an error
+            boolean isValidResponse = isValidResponse(context.getProperty(DuoAuthenticatorConstants.DUO_STATE).
+                    toString(), requestState);
+
+            if (!isValidResponse) {
+                throw new AuthenticationFailedException(DuoAuthenticatorConstants.DuoErrors.ERROR_VERIFY_USER,
+                        "Authentication failed!. Duo response state does not match with the context state");
+            }
+            AuthenticatedUser authenticatedUser = (AuthenticatedUser) context
+                    .getProperty(DuoAuthenticatorConstants.AUTHENTICATED_USER);
+            if (authenticatedUser == null) {
+                throw new AuthenticationFailedException("Authentication failed!. Cannot proceed further without " +
+                        "identifying the user");
+            }
+
+            String username = authenticatedUser.getAuthenticatedSubjectIdentifier();
+
+            if (username == null) {
+                throw new AuthenticationFailedException("Authentication failed!. Cannot proceed further without " +
+                        "identifying the user");
+            }
+            // Step 6: Exchange the auth duoCode for a Token object which contains metadata about authentication.
+            String duoUserId = getDuoUserId(context);
+            Token duoToken = duoClient.exchangeAuthorizationCodeFor2FAResult(requestDuoCode, duoUserId);
```

Comment:
we have to either persist the duoClient or create a new duoClient and get the token. Give preference to creating a new client.